chat-plugins/poll.js

```diff
@@ -0,0 +1,155 @@
+        for (var i in this.options) {
+            output += '<button value="/votepoll ' + (parseInt(i) + 1) + '" name="send">' + (parseInt(i) + 1) + '. ' + this.options[i].name + '</button><br/>';
+        }
+        output += '</div>';
+
+        return output;
+    };
+
+    Poll.prototype.generateResults = function () {
+        var output = '<div class="infobox" style="text-align:center"><p style="font-weight:bold;font-size:14pt">' + this.question + '</p>';
+        for (var i in this.options) {
+            output += '<span style="font-weight:bold;padding:2px">' + (parseInt(i) + 1) + '. ' + this.options[i].name + '</span><br/><span style="font-size:8pt;padding:0px">' + this.options[i].votes + ' votes (' + ((this.options[i].votes * 100) / this.totalVotes) + '% of total).</span><br/>';
+        }
+        output += '</div>';
+
+        return output;
+    };
+
+    Poll.prototype.update = function () {
+        var results = this.generateResults();
+
+        // Update the poll results for everyone that has voted
+        for (var i in this.room.users) {
+            var user = this.room.users[i];
+            if (this.voters.indexOf(user.latestIp) >= 0) {
+                user.sendTo(this.room, '|uhtmlchange|poll|' + results);
+            }
+        }
+    };
+
+    Poll.prototype.display = function (user, broadcast) {
+        var votes = this.generateVotes();
+        var results = this.generateResults();
+
+        var target = [];
+
+        if (broadcast) {
+            target = this.room.users;
+        } else {
+            target.push(user);
+        }
+
+        for (var i in target) {
+            var thisUser = target[i];
+            if (this.voters.indexOf(thisUser.latestIp) >= 0) {
+                thisUser.sendTo(this.room, '|uhtml|poll|' + results);
+            }
+            else {
+                thisUser.sendTo(this.room, '|uhtml|poll|' + votes);
+            }
+        }
+    };
+
+    Poll.prototype.start = function () {
+        var votes = this.generateVotes();
+
+        this.room.send('|uhtml|poll|' + votes);
+    };
+
+    Poll.prototype.end = function () {
+        var results = this.generateResults();
+
+        this.room.send('|uhtml|poll|' + results);
+        this.room.send('|uhtmlchange|poll|' + results);
+    };
+
+    return Poll;
+})();
+
+exports.commands = {
+    newpoll: function (target, room, user) {
+        if (!this.can('ban', null, room)) return false;
+        if (room.poll) return this.sendReply('There is already a poll in progress in this room.');
+
+        var targets = target.split(target.includes('|') ? '|' : ',');
+        if (targets.length < 3) {
+            return this.sendReply('Not enough arguments for /newpoll.');
+        }
+
+        var options = [];
+
+        for (var i = 1; i < targets.length; i++) {
+            options.push(Tools.escapeHTML(targets[i]));
```

Comment:
Escaping should be handled on display, not on parsing imo.

Anyway, it should be done <i>after</i> validating the length of the options array.